storage2/src/state.rs

```diff
@@ -0,0 +1,43 @@
+use std::collections::HashMap;
+
+use jmt::{
+    storage::{TreeReader, TreeWriter},
+    JellyfishMerkleTree,
+};
+
+mod transaction;
+use transaction::Transaction as StateTransaction;
+
+/// State is a lightweight copy-on-write fork of the chain state,
+/// implemented as a RYW cache over a pinned JMT version.
+pub(crate) struct State<'a, R: TreeReader + TreeWriter> {
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    jmt_version: jmt::Version,
+    jmt: &'a JellyfishMerkleTree<'a, R>,
+}
+
+impl<'a, R: TreeReader + TreeWriter> State<'a, R> {
+    pub fn new() -> Self {
+        Self {
+            cache: todo!(),
+            jmt_version: todo!(),
+            jmt: todo!(),
+        }
+    }
+}
+
+impl<'a, R: TreeReader + TreeWriter> StateRead for State<'a, R> {
+    fn get(&self, key: jmt::KeyHash) -> Option<&jmt::OwnedValue> {
+        todo!()
+    }
+}
+
+pub trait StateRead {
+    /// Get
+    fn get(&self, key: jmt::KeyHash) -> Option<&jmt::OwnedValue>;
+}
+
+pub trait StateWrite {
+    /// Copy-on-write put
+    fn put(&mut self, key: jmt::KeyHash, value: jmt::OwnedValue) -> Self;
```

Comment:
I don't think we will ever want to call `put` outside a transaction?

Reply:
@plaidfinch are you suggesting to implement it directly on the `Transaction` type rather than in a trait? I think we might need a `StateWrite` implementation for `State` when the transaction gets applied, or maybe @hdevalence had something else in mind in https://github.com/penumbra-zone/penumbra/issues/1489

Reply:
I don't think we should have a `StateWrite` implementation for `State`; while it would be possible to provide one, that would mean you could put a raw `State` in places where there should always be a transaction. Instead, `State` could have an inherent method of the same signature, if it's needed to implement transaction application?